### PR TITLE
Add DEPRECATIONS.md

### PR DESCRIPTION
**Pull request by zsmb13**
https://stream-io.atlassian.net/browse/CAS-713

### Description

Creates a document listing our deprecated APIs and their expected further deprecation / removal dates.

View it here https://github.com/GetStream/stream-chat-android/blob/feature/deprecations-doc/DEPRECATIONS.md

### CODE REVIEW

File: DEPRECATIONS.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+| `User#unreadCount`<br/>*client* | 2020.11.05.<br/>4.4.2 | 2021.02.22.⌛ | 2021.03.22.⌛ | Use `totalUnreadCount` instead |
+| `ChannelController`<br/>*client* | 2020.11.04.<br/>4.4.1 | 2021.02.22.⌛ | 2021.04.22.⌛ | Renamed due to conflicting name with a type in the offline library, use `ChannelClient` instead |
+| `Chat` interface<br/>*ui-common* | 2020.10.27.<br/>4.4.1 | 2021.02.22.⌛ | 2021.04.22.⌛ | Use `ChatUI` instead |
+| `Pagination#get`<br/>*client* | 2020.10.12.<br/>4.3.0 | 2021.02.22.⌛ | 2021.03.22.⌛ | Use `toString` instead |
+| `MessageListView#setViewHolderFactory`<br/>*ui (old)* | 2020.10.15.<br/>4.3.1 | 2021.02.22.⌛ | 2021.04.22.⌛ | Use the more explicit `setMessageViewHolderFactory` method instead |
+| `MessageListItemAdapter#replaceEntities`<br/>*ui (old)* | - | 2020.10.05.<br/>4.3.0 | 2021.02.22.⌛ | Use `submitList` instead | 
+| ChatObservable based event APIs<br/>*client* | 2020.09.18. | 2021.02.22.⌛ | 2021.04.22⌛ | Replace with direct `subscribe` calls on `ChatClient` and `ChannelClient`, see migration guide (TODO) |
```
Comment (JcMinarro) [inline DEPRECATIONS.md:19]: Maybe add `ChatObservable` as code text?
```suggestion
| `ChatObservable` based event APIs<br/>*client* | 2020.09.18. | 2021.02.22.⌛ | 2021.04.22⌛ | Replace with direct `subscribe` calls on `ChatClient` and `ChannelClient`, see migration guide (TODO) |
```

File: DEPRECATIONS.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# Deprecations
+
+This document lists deprecated constructs in the SDK, with their expected time (⌛) of further deprecations and removals.
+
+| API / Feature | Deprecated (warning) | Deprecated (error) | Removed | Notes |
+| --- | --- | --- | --- | --- |
+| `setUser` (and similar) methods<br/>*client* | 2021.02.03.<br/>4.5.3 | 2021.05.03.⌛ | 2021.08.03.⌛ | Replaced by `connectUser` style methods that return `Call` objects, see migration guide (TODO) |
```
Comment (bychkovdmitry) [inline DEPRECATIONS.md:7]: We should update website docs, I believe `setUser` is still used there for anonymous and guest users

File: DEPRECATIONS.md [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+| `MessageListViewModel.Event.AttachmentDownload`<br/>*ui-common* | 2021.01.29.<br/>4.5.2 | 2021.02.29.⌛ | 2021.03.29.⌛ | Use `DownloadAttachment` instead |
+| `subscribe` methods with Kotlin function parameters<br/>*client* | 2021.01.27.<br/>4.5.2 | 2021.03.27.⌛ | 2021.05.27.⌛ | Use methods with `ChatEventListener` parameters instead (only affects Java clients) |
```
Comment (bychkovdmitry) [inline DEPRECATIONS.md:9]: `2021.01.27.` Do we need dots in the end of every date?

File: DEPRECATIONS.md (hunk 2 above)
Comment (zsmb13) [reply DEPRECATIONS.md:7]: Can you point me to where you still see `setUser` used? Those should be updated already, but if not, it's going on my TODO list.

File: DEPRECATIONS.md (hunk 3 above)
Comment (zsmb13) [reply DEPRECATIONS.md:9]: Probably not, I'll remove them 😄

Comment (JcMinarro) [reply DEPRECATIONS.md]: What about to use `/` or `-` as date separator? :thinking:

Comment (zsmb13) [reply DEPRECATIONS.md]: image

Made me write a regex, but okay 😄

Comment (zsmb13) [reply DEPRECATIONS.md]: image

Not okay, breaks GitHub formatting

Comment (zsmb13) [reply DEPRECATIONS.md]: I don't want to add even more manual `<br/>`s into the doc, editing it is already gonna be painful

File: DEPRECATIONS.md (hunk 2 above)
Comment (bychkovdmitry) [reply DEPRECATIONS.md]: Can't find it anymore, somebody fixed it 👍

**Review by bychkovdmitry — APPROVED** [review]

File: DEPRECATIONS.md [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+| `User#unreadCount`<br/>*client* | 2020.11.05<br/>4.4.2 | 2021.02.22 ⌛ | 2021.03.22 ⌛ | Use `totalUnreadCount` instead |
+| `ChannelController`<br/>*client* | 2020.11.04<br/>4.4.1 | 2021.02.22 ⌛ | 2021.04.22 ⌛ | Renamed due to conflicting name with a type in the offline library, use `ChannelClient` instead |
+| `Chat` interface<br/>*ui-common* | 2020.10.27<br/>4.4.1 | 2021.02.22 ⌛ | 2021.04.22 ⌛ | Use `ChatUI` instead |
+| `Pagination#get`<br/>*client* | 2020.10.12<br/>4.3.0 | 2021.02.22 ⌛ | 2021.03.22 ⌛ | Use `toString` instead |
+| `MessageListView#setViewHolderFactory`<br/>*ui (old)* | 2020.10.15<br/>4.3.1 | 2021.02.22 ⌛ | 2021.04.22 ⌛ | Use the more explicit `setMessageViewHolderFactory` method instead |
+| `MessageListItemAdapter#replaceEntities`<br/>*ui (old)* | - | 2020.10.05<br/>4.3.0 | 2021.02.22 ⌛ | Use `submitList` instead | 
+| `ChatObservable` based event APIs (`events()`)<br/>*client* | 2020.09.18 | 2021.02.22 ⌛ | 2021.04.22 ⌛ | Replace with direct `subscribe` calls on `ChatClient` and `ChannelClient`, see [migration guide](https://github.com/GetStream/stream-chat-android/wiki/Migration-guide:-ChatObserver-and-events()-APIs) |
```
Comment (JcMinarro) [inline DEPRECATIONS.md:19]: Could we update this [Migration guide](https://github.com/GetStream/stream-chat-android/wiki/Migration-guide:-ChatObserver-and-events()-APIs)?
It seems we are recommending using some deprecated methods

**Review by JcMinarro — APPROVED** [review]

File: DEPRECATIONS.md (hunk 4 above)
Comment (zsmb13) [reply DEPRECATIONS.md:19]: The deprecated methods used in that guide are all in the "old API" snippets, aren't they? The guide shows examples of the old API, and how to replace them with the new one.

**Pull request merged**